pandas/core/strings/object_array.py

```diff
@@ -0,0 +1,422 @@
+                return x.get(i)
+            elif len(x) > i >= -len(x):
+                return x[i]
+            return self._default_na_value
+
+        return self._map(f)
+
+    def index(self, sub, start=0, end=None):
+        if end:
+            f = lambda x: x.index(sub, start, end)
+        else:
+            f = lambda x: x.index(sub, start, end)
+        return self._map(f, dtype="int64")
+
+    def rindex(self, sub, start=0, end=None):
+        if end:
+            f = lambda x: x.rindex(sub, start, end)
+        else:
+            f = lambda x: x.rindex(sub, start, end)
+        return self._map(f, dtype="int64")
+
+    def join(self, sep):
+        return self._map(sep.join)
+
+    def partition(self, sep, expand):
+        result = self._map(lambda x: x.partition(sep), dtype="object")
+        return result
+
+    def rpartition(self, sep, expand):
+        return self._map(lambda x: x.rpartition(sep), dtype="object")
+
+    def len(self):
+        return self._map(len, dtype="int64")
+
+    def slice(self, start=None, stop=None, step=None):
+        obj = slice(start, stop, step)
+        return self._map(lambda x: x[obj])
+
+    def slice_replace(self, start=None, stop=None, repl=None):
+        if repl is None:
+            repl = ""
+
+        def f(x):
+            if x[start:stop] == "":
+                local_stop = start
+            else:
+                local_stop = stop
+            y = ""
+            if start is not None:
+                y += x[:start]
+            y += repl
+            if stop is not None:
+                y += x[local_stop:]
+            return y
+
+        return self._map(f)
+
+    def split(self, pat=None, n=-1, expand=False):
+        if pat is None:
+            if n is None or n == 0:
+                n = -1
+            f = lambda x: x.split(pat, n)
+        else:
+            if len(pat) == 1:
+                if n is None or n == 0:
+                    n = -1
+                f = lambda x: x.split(pat, n)
+            else:
+                if n is None or n == -1:
+                    n = 0
+                regex = re.compile(pat)
+                f = lambda x: regex.split(x, maxsplit=n)
+        return self._map(f, dtype=object)
+
+    def rsplit(self, pat=None, n=-1):
+        if n is None or n == 0:
+            n = -1
+        f = lambda x: x.rsplit(pat, n)
+        return self._map(f, dtype="object")
+
+    def translate(self, table):
+        return self._map(lambda x: x.translate(table))
+
+    def wrap(self, width, **kwargs):
+        kwargs["width"] = width
+        tw = textwrap.TextWrapper(**kwargs)
+        return self._map(lambda s: "\n".join(tw.wrap(s)))
+
+    def get_dummies(self, sep="|"):
+        from pandas import Series
+
+        arr = Series(self._array).fillna("")
+        assert isinstance(arr, Series)  # fillna returns Optional[Series]
+        try:
+            arr = sep + arr + sep
+        except TypeError:
+            arr = sep + arr.astype(str) + sep
```

Comment:
@simonjayhawkins any idea why mypy is failing here?

```
pandas/core/strings/object_array.py:348: error: Item "None" of "Optional[Series]" has no attribute "astype"  [union-attr]
```

I would have thought the `assert isinstance(arr, Series)` would make this pass. Declaring `arr: Series = ...` doesn't fix it either.

Reply:
Not sure why. but sep is Any, and `sep + arr + sep` is resolving to Any and arr is reset back to the original type for arr.

`arr = cast(Series, sep + arr + sep)` fixes till we have the binops typed and we can track the cast.

you could also do this instead of the earlier assert so that once Literal is used and the return type of fillna is correct for inplace, we can track the cast.